src/effects/VST3/VST3Effect.h

```diff
@@ -0,0 +1,153 @@
+/**********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  @file VST3Effect.h
+
+  @author Vitaly Sverchinsky
+
+  @brief Part of Audacity VST3 module
+
+**********************************************************************/
+
+
+#pragma once
+
+#include <pluginterfaces/gui/iplugview.h>
+#include <pluginterfaces/vst/ivstaudioprocessor.h>
+#include <public.sdk/source/vst/hosting/module.h>
+
+#include "EffectInterface.h"
+
+#include "SampleCount.h"
+#include "effects/Effect.h"
+#include "internal/ComponentHandler.h"
+
+class NumericTextCtrl;
+
+namespace Steinberg
+{
+   namespace Vst
+   {
+      class IComponent;
+      class IEditController;
+      class IParameterChanges;
+      class IComponentHandler;
+      class IConnectionPoint;
+   }
+}
+
+class ParameterChangesProvider;
+
+/**
+ * \brief Objects of this class connect Audacity with VST3 effects
+ */
+class VST3Effect final : public EffectUIClientInterface
+{
+   //Keep strong reference to a module while effect is alive
+   std::shared_ptr<VST3::Hosting::Module> mModule;
+
+   //Following fields are unique to each effect instance
+
+   Steinberg::IPtr<Steinberg::Vst::IComponent> mEffectComponent;
+   Steinberg::IPtr<Steinberg::Vst::IAudioProcessor> mAudioProcessor;
+   Steinberg::IPtr<Steinberg::Vst::IConnectionPoint> mComponentConnectionProxy;
+   Steinberg::IPtr<Steinberg::Vst::IConnectionPoint> mControllerConnectionProxy;
+   Steinberg::Vst::ProcessSetup mSetup;
+   const VST3::Hosting::ClassInfo mEffectClassInfo;
+
+   //Since all of the realtime processors share same presets, following
+   //fields are only initialized and assigned in the global effect instance
+
+   Steinberg::IPtr<Steinberg::IPlugView> mPlugView;
+   Steinberg::IPtr<Steinberg::Vst::IEditController> mEditController;
+   Steinberg::IPtr<internal::ComponentHandler> mComponentHandler;
```

Comment:
The non-copy constructor may leave mEditController null.  It makes mComponentHandler non-null only if mEditController is.  These things should be stated here in comments.

In my first round of review, I looked at uses of these two fields to see if null checks were sufficient.

Every use of mEditController should be preceded by a null check of it or of mComponentHandler.  That condition is satisfied everywhere.

Every use of mComponentHandler needs a null check.

I checked again and it looks good now.